### PR TITLE
feat: implement blockdevice watch controller

### PR DESCRIPTION
**Pull request by smira**
This controller combines kobject events, and scan of `/sys/block` to build a consistent list of available block devices, updating resources as the blockdevice changes.

Based on these resources the next step can run probe on the blockdevices as they change to present a consistent view of filesystems/partitions.

Part of #8367

### CODE REVIEW

File: internal/app/machined/pkg/controllers/block/devices.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	}
+}
+
+func (ctrl *DevicesController) resync(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	events, err := sysblock.Walk("/sys/block")
+	if err != nil {
+		return fmt.Errorf("failed to walk /sys/block: %w", err)
+	}
+
+	touchedIDs := make(map[string]struct{}, len(events))
```
Comment (frezbo) [inline internal/app/machined/pkg/controllers/block/devices.go:104]: `touchedIDs` supposed to be scoped to `resync` lifetime?

File: internal/app/machined/pkg/controllers/block/devices.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+		zap.String("path", ev.DevicePath),
+		zap.String("id", ev.Values["DEVNAME"]),
+	)
+
+	logger.Debug("processing event")
+
+	id := ev.Values["DEVNAME"]
+
+	// re-stat the sysfs entry to make sure we are not of sync with events
+	_, err := os.Stat(ev.DevicePath)
```
Comment (frezbo) [inline internal/app/machined/pkg/controllers/block/devices.go:147]: `err` not checked?

File: internal/app/machined/pkg/controllers/block/devices.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+			dev.TypedSpec().DevicePath = ev.DevicePath
+
+			dev.TypedSpec().Generation++
+
+			return nil
+		}); err != nil {
+			return fmt.Errorf("failed to modify device %q: %w", id, err)
+		}
+	case kobject.ActionRemove:
+		if err == nil { // entry still exists, skip removing
```
Comment (frezbo) [inline internal/app/machined/pkg/controllers/block/devices.go:173]: hmm maybe call this `statErr` so the intent is clear

File: internal/app/machined/pkg/controllers/block/internal/sysblock/sysblock.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			return nil, err
+		}
+
+		result = append(result, &kobject.Event{
+			Action:     kobject.Add,
```
Comment (frezbo) [inline internal/app/machined/pkg/controllers/block/internal/sysblock/sysblock.go:61]: is this always an `Add` event?

File: internal/app/machined/pkg/controllers/block/devices.go (hunk 1 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/block/devices.go:104]: yes

File: internal/app/machined/pkg/controllers/block/internal/sysblock/sysblock.go (hunk 4 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/block/internal/sysblock/sysblock.go:61]: yes, as we're injecting current contents

File: internal/app/machined/pkg/controllers/block/devices.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		zap.String("action", string(ev.Action)),
+		zap.String("path", ev.DevicePath),
+		zap.String("id", ev.Values["DEVNAME"]),
+	)
+
+	logger.Debug("processing event")
+
+	id := ev.Values["DEVNAME"]
+
+	// re-stat the sysfs entry to make sure we are not of sync with events
```
[the remark this answers is not in the document]
Comment (smira) [reply internal/app/machined/pkg/controllers/block/devices.go]: thanks, fixed

**Review by frezbo — APPROVED** [review]

**Comment by smira** [comment]
Added an integration test.

Screencast of live updated:

[Screencast from 2024-03-14 19-58-52.webm](https://github.com/siderolabs/talos/assets/55211/7aa345b5-d3ea-4774-aca6-0236649fdf67)

File: internal/integration/api/volumes.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				continue
+			}
+		}
+
+		suite.Assert().Equal(expected.Name, dv.TypedSpec().Name)
+
+		delete(expectedVolumes, id)
+	}
+
+	suite.Assert().Empty(expectedVolumes)
```
Comment (frezbo) [inline internal/integration/api/volumes.go:95]: this is nice, we could use the blockdevices in extensions test, currently it's a hacky way to discover free disks

**Pull request merged**